### PR TITLE
Add DQfD

### PR DESCRIPTION
**Pull request by MrSyee**
Squashed commit of the following:

commit e458ec6fea48fd1e1deabf6bb2c61f55737d26bf
Author: khkim <kh.kim@medipixel.io>
Date:   Thu Feb 28 22:57:32 2019 +0900

    Add dqfd

commit 01bbc929a5eda576691d9bc028a629e0e3eb0799
Author: khkim <kh.kim@medipixel.io>
Date:   Thu Feb 28 17:23:22 2019 +0900

    Change perfd buffer to inherit per buffer

    Squashed commit of the following:

    commit 5512696d8fe9ef8cfcc5639daea2b83a2299ebf0
    Author: khkim <kh.kim@medipixel.io>
    Date:   Thu Feb 28 16:54:01 2019 +0900

        Change perfd buffer to inherit per buffer

    commit d7043350612757f8e595ee72de916f0f5a5269bf
    Author: khkim <kh.kim@medipixel.io>
    Date:   Thu Feb 28 16:50:18 2019 +0900

        Change numpy version

    commit 3e5155089ed6ec8bb14bdc1f0cbd0a237ae686ca
    Author: khkim <kh.kim@medipixel.io>
    Date:   Thu Feb 28 16:10:46 2019 +0900

        Change .travis

    commit 6a81955b0128a3bb36d62d22aef5908505b96009
    Author: khkim <kh.kim@medipixel.io>
    Date:   Thu Feb 28 15:42:18 2019 +0900

        Change requirements-dev

    commit 7d5e29fa720bb5e0aad0bace9c8075ad347e369d
    Author: khkim <kh.kim@medipixel.io>
    Date:   Thu Feb 28 15:24:52 2019 +0900

        Change requirements-dev

    commit ac9dabb4c68121e52a46e38f7ebdb2451b31ef39
    Author: khkim <kh.kim@medipixel.io>
    Date:   Thu Feb 28 15:14:23 2019 +0900

        Change .isort.cfg

    commit 2bca8459e354be934e8afe27115fbf4e795f1c95
    Author: khkim <kh.kim@medipixel.io>
    Date:   Thu Feb 28 14:51:19 2019 +0900

        Modify requirements

    commit bf9408b181779549dbc18866052ab905b5b8da65
    Author: khkim <kh.kim@medipixel.io>
    Date:   Thu Feb 28 14:26:19 2019 +0900

        Modify requirements-dev

    commit 732ac6f79958f59fd94e0390fe088b69c76a6dbc
    Author: khkim <kh.kim@medipixel.io>
    Date:   Thu Feb 28 14:16:33 2019 +0900

        Modify requirements-dev

    commit 2b8a6867f97c5350f7e77a45f0e514e2ca65f708
    Author: khkim <kh.kim@medipixel.io>
    Date:   Thu Feb 28 14:15:18 2019 +0900

        Modify requirements-dev and trivis.yml

    commit 0c380a131a1372764b2b093d0aeef35dc30b651d
    Author: khkim <kh.kim@medipixel.io>
    Date:   Thu Feb 28 13:54:37 2019 +0900

        Modify requirements-dev

    commit c1d538da551a9afd0e15f76c06eef0622a16291a
    Author: khkim <kh.kim@medipixel.io>
    Date:   Thu Feb 28 13:49:03 2019 +0900

        Modify requirements-dev

    commit 5695cd82f915eff89f9eb59a11b33aa379e9e9d1
    Author: khkim <kh.kim@medipixel.io>
    Date:   Thu Feb 28 13:21:10 2019 +0900

        Modify requirements-dev

    commit e5c923604de83a0e8bbf55d58f53159ee92e548b
    Author: khkim <kh.kim@medipixel.io>
    Date:   Thu Feb 28 12:20:29 2019 +0900

        Modify pytest-pylint version on requirements-dev

    commit b2f68ac2beba0b58603f7fb027d4b135ec63bc81
    Author: khkim <kh.kim@medipixel.io>
    Date:   Thu Feb 28 12:13:30 2019 +0900

        Modify pylint version on requirements-dev

    commit 4c8556e1419aca4a100f690f3449327c9f1a1e4e
    Author: khkim <kh.kim@medipixel.io>
    Date:   Thu Feb 28 12:02:47 2019 +0900

        Input epsilon when using DQN with dueling

    commit d26764fbc2bab428b0077f64c14b659442c300f5
    Author: khkim <kh.kim@medipixel.io>
    Date:   Thu Feb 28 11:49:08 2019 +0900

        Add dueling on DQN

        Squashed commit of the following:

        commit ee459c3b8b1c448abaa4373d798879250867652b
        Author: khkim <kh.kim@medipixel.io>
        Date:   Fri Feb 22 20:42:14 2019 +0900

            Add dueling network with q value calculating

        commit 49d261fbe66782ff1de239a348e690337fd67517
        Author: khkim <kh.kim@medipixel.io>
        Date:   Fri Feb 22 17:36:00 2019 +0900

            Modify dueling output method

        commit 1f79e7ca63114742180da9f42b133c9e2bf8153a
        Author: khkim <kh.kim@medipixel.io>
        Date:   Fri Feb 22 15:00:00 2019 +0900

            Add dueling network for dqn

            Squashed commit of the following:

            commit bce3fe01262d436a6845207417434f0e68be9b1a
            Author: khkim <kh.kim@medipixel.io>
            Date:   Fri Feb 22 11:37:51 2019 +0900

                Test dueling and add regularization term

            commit f673e442a9c877529c8fca80c727d5e7889c3253
            Author: khkim <kh.kim@medipixel.io>
            Date:   Thu Feb 21 11:38:25 2019 +0900

                Add dqn

                Squashed commit of the following:

                commit 011e5640e2a4cf64de6121706a96ddba55218b1e
                Author: khkim <kh.kim@medipixel.io>
                Date:   Thu Feb 21 10:23:49 2019 +0900

                    Add dqn

                    Squashed commit of the following:

                    commit 545f63b22c3514336173d7df18b67ed33a55c664
                    Author: khkim <kh.kim@medipixel.io>
                    Date:   Tue Feb 19 20:26:56 2019 +0900

                        Modify dqn/net to dqn/networks

                    commit 1c8f861315c8010984acfa639246ac9409ea6926
                    Author: khkim <kh.kim@medipixel.io>
                    Date:   Tue Feb 19 20:18:44 2019 +0900

                        Merge DQN in discrete env

                        Squashed commit of the following:

                        commit 6906a756cd367c5c0a2c8576c19a70d03cab7ebc
                        Author: khkim <kh.kim@medipixel.io>
                        Date:   Tue Feb 19 18:24:04 2019 +0900

                            Modify DQN

                        commit c45ab325089d24a492355cf29b80ffa315bf258d
                        Author: khkim <kh.kim@medipixel.io>
                        Date:   Mon Feb 18 17:12:56 2019 +0900

                            Add DQN on lunarlander_v2

### CODE REVIEW

File: .isort.cfg [hunk 1]
```diff
@@ -5,3 +5,4 @@ force_grid_wrap=0
 combine_as_imports=True
 line_length=88
 force_sort_within_sections=True
+known_third_party=wandb
```
Comment (MrSyee) [inline .isort.cfg]: wandb import가 계속해서 바뀌는 문제 때문에 고정시킴

File: algorithms/common/buffer/priortized_replay_buffer.py [hunk 2]
```diff
@@ -137,24 +137,18 @@ def update_priorities(self, indices: list, priorities: np.ndarray):
             self._max_priority = max(self._max_priority, priority)
 
 
-class PrioritizedReplayBufferfD(ReplayBuffer):
+class PrioritizedReplayBufferfD(PrioritizedReplayBuffer):
```
Comment (MrSyee) [inline algorithms/common/buffer/priortized_replay_buffer.py]: PrioritizedReplayBufferfD class 가 PrioritizedReplayBuffer를 상속받도록 변경

File: algorithms/fd/dqn_agent.py [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+        )
+
+        # G_t   = r + gamma * v(s_{t+1})  if state != Terminal
+        #       = r                       otherwise
+        masks = 1 - dones
+
+        target = rewards + self.hyper_params["GAMMA"] * next_q_value * masks
+        target = target.to(device)
+
+        # supervised loss using demo - when pretrain
```
Comment (MrSyee) [inline algorithms/fd/dqn_agent.py]: 기존 ddpgfd와 다른 loss 추가
논문의 3page loss function 참조 (https://arxiv.org/pdf/1704.03732.pdf)

File: examples/lunarlander_v2/dqfd.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    "PER_BETA": 0.4,
+    "PER_EPS": 1e-6,
+    "UPDATE_STARTS_FROM": int(1e4),
+    "MULTIPLE_LEARN": n_cpu // 2 if n_cpu >= 2 else 1,
+    "N_WORKERS": n_cpu,
+    "PRETRAIN_STEP": 100,
+    "LAMDA1": 1.0,  # N-step return weight
+    "LAMDA2": 1.0,  # Supervised loss weight
+    "LAMDA3": 1e-5,  # l2 regularization weight
+    "MARGIN": 0.8,  # margin for supervised loss
```
Comment (MrSyee) [inline examples/lunarlander_v2/dqfd.py]: supervised loss 에 필요한 margin 값 추가

File: examples/lunarlander_v2/dqfd.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    "PER_ALPHA": 0.5,
+    "PER_BETA": 0.4,
+    "PER_EPS": 1e-6,
+    "UPDATE_STARTS_FROM": int(1e4),
+    "MULTIPLE_LEARN": n_cpu // 2 if n_cpu >= 2 else 1,
+    "N_WORKERS": n_cpu,
+    "PRETRAIN_STEP": 100,
+    "LAMDA1": 1.0,  # N-step return weight
+    "LAMDA2": 1.0,  # Supervised loss weight
+    "LAMDA3": 1e-5,  # l2 regularization weight
```
Comment (MrSyee) [inline examples/lunarlander_v2/dqfd.py]: WEIGHT_DECAY를 논문의 기호인 LAMDA3로 변경

File: run_lunarlander_v2.py [hunk 6]
```diff
@@ -40,6 +40,12 @@
 parser.add_argument(
     "--max-episode-steps", type=int, default=300, help="max episode step"
 )
+parser.add_argument(
```
Comment (MrSyee) [inline run_lunarlander_v2.py]: 데모 path 추가

**Comment by Curt-Park** [comment]
고생하셨습니다. 약간의 튜닝작업 이후 머지하겠습니다.

File: algorithms/fd/dqn_agent.py (hunk 3 above)
Comment (Curt-Park) [reply algorithms/fd/dqn_agent.py]: 1. pred_actions = torch.argmax(q_value + margin_value)가 되어야하지 않나요?
2. margin[np.where(actions != pred_actions.float())] = margin_value 에서 float으로 !=연산을 하게되면 오차가 발생할 가능성이 있지 않을까요?

Comment (Curt-Park) [reply algorithms/fd/dqn_agent.py]: 3. if np.sum(eps_d) == self.hyper_params["BATCH_SIZE"]:
pre-training 시에만 돌린다는 의미인건가요? eps_d가 1이 아닌 값으로 세팅되면 어떻게 되나요?
더불어 np.sum(eps_d)은 float이고, self.hyper_params["BATCH_SIZE"]는 int입니다.

Comment (Curt-Park) [reply algorithms/fd/dqn_agent.py]: 4. 논문에 의하면 아래 loss는 pre-training에만 적용되는건 아닌것 같습니다.
```All the losses are applied to the demonstration data in both phases, while the supervised loss is not
applied to self-generated data (λ2 = 0)```

File: algorithms/fd/dqn_agent.py [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+        # update priorities in PER
+        new_priorities = (target - curr_q_value).pow(2)
+        new_priorities = (
+            new_priorities.data.cpu().numpy().squeeze() + self.hyper_params["PER_EPS"]
+        )
+        new_priorities += eps_d
+        self.memory.update_priorities(indexes, new_priorities)
+
+        # decrease epsilon
+        max_epsilon, min_epsilon, epsilon_decay = (
```
Comment (Curt-Park) [inline algorithms/fd/dqn_agent.py]: Decrease epsilon 코드가 두 번 들어가있네요. 이 부분은 tuned parameter update와 함께 수정하겠습니다.

Comment (Curt-Park) [reply algorithms/fd/dqn_agent.py]: 수정완료

File: algorithms/fd/dqn_agent.py (hunk 3 above)
Comment (MrSyee) [reply algorithms/fd/dqn_agent.py]: 1. margin을 구할때 현재 action과 데모 action을 비교해야해서 state가 주어졌을때 greedy policy라 생각하고 argmax하였습니다. 지금 생각해보니 select_action 함수를 적용해야하나 싶기도 하네요.
2. 비교하는 두개의 action이 타입이 안맞아 에러가 나서 맞춘것이었는데 actions를 long으로 바꾸는 것이 더 맞을 것 같네요.
3. 4. 네. pre-training시만 돌리게 하려고 조건문 걸어 논 것입니다. 모든 경우에 대해서 loss를 적용하려고 했는데 시간이 부족하여 일단 loss가 적용되는지만 확인하려고 적어두었습니다. 제가 깜빡하고 이부분에 대해서 코멘트를 못달아 드렸네요 ㅠ. 모든 학습시 이 loss가 적용되도록 해야할 것 같습니다. 저 loss는 데모 데이터를 사용할때만 사용할 수 있는데 현재 batch 뽑은 것 중에 데모 데이터인것만 체크해서 저 loss를 적용하는 로직이 필요할 것 같습니다.

**Comment by MrSyee** [comment]
supervised loss에 대한 코멘트 추가했습니다. single comment로 달았더니 메일이 안가네요.

https://github.com/medipixel/rl_algorithms/pull/100/files/0adc2450401c39168adebe4d33ad8a4bce7b82b6#r261208855

2019년 3월 1일 (금) 오후 5:30, Jinwoo Park (Curt) <notifications@github.com>님이 작성:

File: algorithms/fd/dqn_agent.py (hunk 3 above)
Comment (Curt-Park) [reply algorithms/fd/dqn_agent.py]: @MrSyee 첫번째 코멘트는 논문의 수식을 보고 드렸습니다.
논문의 수식과 같아지려면 
`pred_actions = torch.argmax(q_value + margin_value)`
가 되어야 합니다. 이렇게 하는 것이 논문의 취지와 맞는것 같기도 하구요. (state에 부합되는 demo action외의 action들에 대해서는 value를 낮추기 위함)

**Comment by Curt-Park** [comment]
dqfd_exp

튜닝완료. 지상 최강의 로켓 학습자 완성.

1. supervised loss는 priority 계산에 사용되면 성능이 하락됨
2. deformable object manipulator의 loss idea (eps_d에 loss를 곱)를 적용해보았으나 성능향상 x
3. 학습이 불안정한 것은 gradient clipping을 사용하여 잡음
